### ISSUE TITLE
Encountered filters that are not defined in the schema

### ISSUE THREAD

**Issue by Albvadi** [laravel-json-api/laravel#115]
Hello there,
First of all, thank you for this amazing package. It is difficult to find a package that facilitates so much work and standardizes everything in such a simple way!

I have a problem with my filters. In one of the resources, I try to filter on the field of a relation (defined in its own schema), but it tells me all the time that the defined filter was not found.
This is my code:

Recurso Eloquent Model
```php
class Recurso extends Model
{
    public function municipio()
    {
        return $this->belongsTo(Municipio::class, 'municipio_id', 'id');
    }
}
```
RecursoSchema JSON:API Schema
```php
class RecursoSchema extends Schema
{
    public static string $model = Recurso::class;

    protected int $maxDepth = 4;

    public function fields(): array
    {
        return [
            ID::make(),
            Str::make('nombre')->sortable()->readOnly(),
            Str::make('descripcion')->sortable()->readOnly(),
            Str::make('direccion')->sortable()->readOnly(),
            Str::make('telefono')->sortable()->readOnly(),
            Str::make('movil')->sortable()->readOnly(),
            Str::make('email')->sortable()->readOnly(),
            Str::make('coordenadas')->sortable()->readOnly(),
            BelongsTo::make('municipio')->type('municipios')
            DateTime::make('createdAt')->sortable()->readOnly(),
            DateTime::make('updatedAt')->sortable()->readOnly(),
        ];
    }

    public function filters(): array
    {
        return [
            WhereIdIn::make($this)->delimiter(','),
            Where::make('nombre'),
            WhereHas::make($this, 'municipio'),
        ];
    }
}
```

MunicipioSchema JSON:API Schema
```php
class MunicipioSchema extends Schema
{
    public static string $model = Municipio::class;

    protected int $maxDepth = 4;

    public function fields(): array
    {
        return [
            ID::make()->sortable(),
            Str::make('nombre')->sortable()->readOnly(),
            Str::make('codigo_postal')->readOnly(),
            Str::make('slug')->readOnly(),
            BelongsTo::make('provincia')->type('provincias'),
            DateTime::make('createdAt')->sortable()->readOnly(),
            DateTime::make('updatedAt')->sortable()->readOnly(),
        ];
    }

    public function filters(): array
    {
        return [
            Where::make('nombre'),
            Where::make('slug'),
        ];
    }
}
```

With this code:
  - `/recursos?filter[municipio][nombre]=valverde`: returns all the `recursos` belongs to a `municipio` with that `nombre`.
  - `/recursos?filter[municipio][slug]=valverde`: return error `"Encountered filters that are not defined on the recursos schema or municipio relation: slug"`

At this point, I have different results based on the code changes I make:
  - If I remove `Where::make('nombre'),` from `MunicipioSchema`, I still get the results explained above. I don´t know why `nombre` filter still working.. 🤔 
  - If I change `BelongsTo::make('municipio')->type('municipios')` to `BelongsTo::make('municipio')->type('municipios')->withFilters(
                Where::make('nombre'),
                Where::make('slug'),
            ),` both filters `nombre` and `slug` works as expected.

Am I forgetting to configure something?

Thank you!

**Comment by lindyhopchris** [comment]
Yeah not sure is my initial answer. I haven't used these new relationship filters yet, and the PR that was submitted for them didn't include any tests... so it's possible there's a bug.

Can you share a stack trace for the error "Encountered filters...."?

**Comment by lindyhopchris** [comment]
I've labelled as a bug because I *suspect* this is a bug... but not 100% sure at the mo until I investigate in more detail.

**Comment by Albvadi** [comment]
Of course, this is the stacktrace:
```log
RuntimeException: Encountered filters that are not defined on the recursos schema or municipio relation: slug in file C:\laragon\www\pcomunitaria\vendor\laravel-json-api\eloquent\src\QueryBuilder\Applicators\FilterApplicator.php on line 157

#0 C:\laragon\www\pcomunitaria\vendor\laravel-json-api\eloquent\src\QueryBuilder\Applicators\FilterApplicator.php(113): LaravelJsonApi\Eloquent\QueryBuilder\Applicators\FilterApplicator->rejectUnrecognised(Object(LaravelJsonApi\Core\Query\FilterParameters), Array)
#1 C:\laragon\www\pcomunitaria\vendor\laravel-json-api\eloquent\src\Filters\WhereHas.php(84): LaravelJsonApi\Eloquent\QueryBuilder\Applicators\FilterApplicator->apply(Object(Illuminate\Database\Eloquent\Builder), Object(LaravelJsonApi\Core\Query\FilterParameters))
#2 C:\laragon\www\pcomunitaria\vendor\laravel\framework\src\Illuminate\Database\Eloquent\Builder.php(1174): LaravelJsonApi\Eloquent\Filters\WhereHas->LaravelJsonApi\Eloquent\Filters\{closure}(Object(Illuminate\Database\Eloquent\Builder))
#3 C:\laragon\www\pcomunitaria\vendor\laravel\framework\src\Illuminate\Database\Eloquent\Concerns\QueriesRelationships.php(56): Illuminate\Database\Eloquent\Builder->callScope(Object(Closure))
#4 C:\laragon\www\pcomunitaria\vendor\laravel\framework\src\Illuminate\Database\Eloquent\Concerns\QueriesRelationships.php(147): Illuminate\Database\Eloquent\Builder->has(Object(Illuminate\Database\Eloquent\Relations\BelongsTo), '>=', 1, 'and', Object(Closure))
#5 C:\laragon\www\pcomunitaria\vendor\laravel-json-api\eloquent\src\Filters\WhereHas.php(70): Illuminate\Database\Eloquent\Builder->whereHas('municipio', Object(Closure))
#6 C:\laragon\www\pcomunitaria\vendor\laravel-json-api\eloquent\src\QueryBuilder\Applicators\FilterApplicator.php(98): LaravelJsonApi\Eloquent\Filters\WhereHas->apply(Object(Illuminate\Database\Eloquent\Builder), Array)
#7 C:\laragon\www\pcomunitaria\vendor\laravel-json-api\eloquent\src\QueryBuilder\JsonApiBuilder.php(164): LaravelJsonApi\Eloquent\QueryBuilder\Applicators\FilterApplicator->apply(Object(Illuminate\Database\Eloquent\Builder), Object(LaravelJsonApi\Core\Query\FilterParameters))
#8 C:\laragon\www\pcomunitaria\vendor\laravel-json-api\eloquent\src\QueryBuilder\JsonApiBuilder.php(147): LaravelJsonApi\Eloquent\QueryBuilder\JsonApiBuilder->filter(Object(LaravelJsonApi\Core\Query\FilterParameters))
#9 C:\laragon\www\pcomunitaria\vendor\laravel-json-api\eloquent\src\QueryAll.php(98): LaravelJsonApi\Eloquent\QueryBuilder\JsonApiBuilder->withQueryParameters(Object(LaravelJsonApi\Core\Query\Custom\ExtendedQueryParameters))
#10 C:\laragon\www\pcomunitaria\vendor\laravel-json-api\eloquent\src\QueryAll.php(192): LaravelJsonApi\Eloquent\QueryAll->query()
#11 C:\laragon\www\pcomunitaria\vendor\laravel-json-api\core\src\Core\Store\QueryAllHandler.php(45): LaravelJsonApi\Eloquent\QueryAll->firstOrPaginate(NULL)
#12 C:\laragon\www\pcomunitaria\vendor\laravel-json-api\laravel\src\Http\Controllers\Actions\FetchMany.php(58): LaravelJsonApi\Core\Store\QueryAllHandler->firstOrPaginate(NULL)
#13 C:\laragon\www\pcomunitaria\vendor\laravel\framework\src\Illuminate\Routing\Controller.php(54): App\Http\Controllers\Api\V1\MunicipioController->index(Object(LaravelJsonApi\Laravel\Routing\Route), Object(LaravelJsonApi\Core\Store\Store), 'recursos')
#14 C:\laragon\www\pcomunitaria\vendor\laravel\framework\src\Illuminate\Routing\ControllerDispatcher.php(45): Illuminate\Routing\Controller->callAction('index', Array)
#15 C:\laragon\www\pcomunitaria\vendor\laravel\framework\src\Illuminate\Routing\Route.php(254): Illuminate\Routing\ControllerDispatcher->dispatch(Object(Illuminate\Routing\Route), Object(App\Http\Controllers\Api\V1\MunicipioController), 'index')
#16 C:\laragon\www\pcomunitaria\vendor\laravel\framework\src\Illuminate\Routing\Route.php(197): Illuminate\Routing\Route->runController()
#17 C:\laragon\www\pcomunitaria\vendor\laravel\framework\src\Illuminate\Routing\Router.php(695): Illuminate\Routing\Route->run()
#18 C:\laragon\www\pcomunitaria\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php(128): Illuminate\Routing\Router->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#19 C:\laragon\www\pcomunitaria\vendor\laravel-json-api\laravel\src\Http\Middleware\BootJsonApi.php(103): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#20 C:\laragon\www\pcomunitaria\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php(167): LaravelJsonApi\Laravel\Http\Middleware\BootJsonApi->handle(Object(Illuminate\Http\Request), Object(Closure), 'v1')
#21 C:\laragon\www\pcomunitaria\vendor\laravel\framework\src\Illuminate\Routing\Middleware\ThrottleRequests.php(127): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#22 C:\laragon\www\pcomunitaria\vendor\laravel\framework\src\Illuminate\Routing\Middleware\ThrottleRequests.php(103): Illuminate\Routing\Middleware\ThrottleRequests->handleRequest(Object(Illuminate\Http\Request), Object(Closure), Array)
#23 C:\laragon\www\pcomunitaria\vendor\laravel\framework\src\Illuminate\Routing\Middleware\ThrottleRequests.php(55): Illuminate\Routing\Middleware\ThrottleRequests->handleRequestUsingNamedLimiter(Object(Illuminate\Http\Request), Object(Closure), 'api', Object(Closure))
#24 C:\laragon\www\pcomunitaria\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php(167): Illuminate\Routing\Middleware\ThrottleRequests->handle(Object(Illuminate\Http\Request), Object(Closure), 'api')
#25 C:\laragon\www\pcomunitaria\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php(103): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#26 C:\laragon\www\pcomunitaria\vendor\laravel\framework\src\Illuminate\Routing\Router.php(697): Illuminate\Pipeline\Pipeline->then(Object(Closure))
#27 C:\laragon\www\pcomunitaria\vendor\laravel\framework\src\Illuminate\Routing\Router.php(672): Illuminate\Routing\Router->runRouteWithinStack(Object(Illuminate\Routing\Route), Object(Illuminate\Http\Request))
#28 C:\laragon\www\pcomunitaria\vendor\laravel\framework\src\Illuminate\Routing\Router.php(636): Illuminate\Routing\Router->runRoute(Object(Illuminate\Http\Request), Object(Illuminate\Routing\Route))
#29 C:\laragon\www\pcomunitaria\vendor\laravel\framework\src\Illuminate\Routing\Router.php(625): Illuminate\Routing\Router->dispatchToRoute(Object(Illuminate\Http\Request))
#30 C:\laragon\www\pcomunitaria\vendor\laravel\framework\src\Illuminate\Foundation\Http\Kernel.php(166): Illuminate\Routing\Router->dispatch(Object(Illuminate\Http\Request))
#31 C:\laragon\www\pcomunitaria\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php(128): Illuminate\Foundation\Http\Kernel->Illuminate\Foundation\Http\{closure}(Object(Illuminate\Http\Request))
#32 C:\laragon\www\pcomunitaria\vendor\laravel\framework\src\Illuminate\Foundation\Http\Middleware\TransformsRequest.php(21): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#33 C:\laragon\www\pcomunitaria\vendor\laravel\framework\src\Illuminate\Foundation\Http\Middleware\ConvertEmptyStringsToNull.php(31): Illuminate\Foundation\Http\Middleware\TransformsRequest->handle(Object(Illuminate\Http\Request), Object(Closure))
#34 C:\laragon\www\pcomunitaria\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php(167): Illuminate\Foundation\Http\Middleware\ConvertEmptyStringsToNull->handle(Object(Illuminate\Http\Request), Object(Closure))
#35 C:\laragon\www\pcomunitaria\vendor\laravel\framework\src\Illuminate\Foundation\Http\Middleware\TransformsRequest.php(21): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#36 C:\laragon\www\pcomunitaria\vendor\laravel\framework\src\Illuminate\Foundation\Http\Middleware\TrimStrings.php(40): Illuminate\Foundation\Http\Middleware\TransformsRequest->handle(Object(Illuminate\Http\Request), Object(Closure))
#37 C:\laragon\www\pcomunitaria\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php(167): Illuminate\Foundation\Http\Middleware\TrimStrings->handle(Object(Illuminate\Http\Request), Object(Closure))
#38 C:\laragon\www\pcomunitaria\vendor\laravel\framework\src\Illuminate\Foundation\Http\Middleware\ValidatePostSize.php(27): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#39 C:\laragon\www\pcomunitaria\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php(167): Illuminate\Foundation\Http\Middleware\ValidatePostSize->handle(Object(Illuminate\Http\Request), Object(Closure))
#40 C:\laragon\www\pcomunitaria\vendor\laravel\framework\src\Illuminate\Foundation\Http\Middleware\PreventRequestsDuringMaintenance.php(86): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#41 C:\laragon\www\pcomunitaria\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php(167): Illuminate\Foundation\Http\Middleware\PreventRequestsDuringMaintenance->handle(Object(Illuminate\Http\Request), Object(Closure))
#42 C:\laragon\www\pcomunitaria\vendor\fruitcake\laravel-cors\src\HandleCors.php(52): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#43 C:\laragon\www\pcomunitaria\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php(167): Fruitcake\Cors\HandleCors->handle(Object(Illuminate\Http\Request), Object(Closure))
#44 C:\laragon\www\pcomunitaria\vendor\fideloper\proxy\src\TrustProxies.php(57): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#45 C:\laragon\www\pcomunitaria\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php(167): Fideloper\Proxy\TrustProxies->handle(Object(Illuminate\Http\Request), Object(Closure))
#46 C:\laragon\www\pcomunitaria\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php(103): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#47 C:\laragon\www\pcomunitaria\vendor\laravel\framework\src\Illuminate\Foundation\Http\Kernel.php(141): Illuminate\Pipeline\Pipeline->then(Object(Closure))
#48 C:\laragon\www\pcomunitaria\vendor\laravel\framework\src\Illuminate\Foundation\Http\Kernel.php(110): Illuminate\Foundation\Http\Kernel->sendRequestThroughRouter(Object(Illuminate\Http\Request))
#49 C:\laragon\www\pcomunitaria\public\index.php(52): Illuminate\Foundation\Http\Kernel->handle(Object(Illuminate\Http\Request))
#50 {main}
```

I added dump in the L157 of `FilterApplicator.php` and this is the result:
```php
FilterParameters $filters

LaravelJsonApi\Core\Query\FilterParameters {#1454
  -stack: array:1 [
    "slug" => LaravelJsonApi\Core\Query\FilterParameter {#1455
      -key: "slug"
      -value: "valverde"
    }
  ]
}
```
```php
$allowedFilterKeys

array:3 [
  0 => "id"
  1 => "nombre"
  2 => "municipio"
]
```
I think I have another "bug" searching in a nested relation (same error), but I´ll wait to resolve this and see if the another bug it´s resolved too.

Thank you so much for your quick response!

**Comment by lindyhopchris** [comment]
I believe this has been fixed as a result of #139 being fixed - will be in `v1.0.1`. If the problem persists, please reopen the issue.
